From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nds-ports <nds-ports@localhost>
Date: Tue, 6 Oct 2026 12:00:00 +0200
Subject: [PATCH] Use single precision and reduce floating point work

Use float math for trig, projection and OPL tables. Share the direction
conversion and multiply by a constant instead of dividing for each aim.
---

diff --git a/wl_def.h b/wl_def.h
--- a/wl_def.h
+++ b/wl_def.h
@@ -183,7 +183,7 @@
 
 #define RUNSPEED        6000
 
-#define HEIGHTRATIO     0.50
+#define HEIGHTRATIO     0.50f
 
 // The DS top screen shows the view at 1:1 in the top left corner of the
 // screen buffer, menus and pictures are shown scaled
@@ -206,7 +206,7 @@
                                         // to any actor center
 
 #undef M_PI
-#define PI              3.141592657
+#define PI              3.141592657f
 #define M_PI PI
 
 #define GLOBAL1         (1l<<16)
diff --git a/wl_main.c b/wl_main.c
--- a/wl_main.c
+++ b/wl_main.c
@@ -719,7 +719,7 @@
     int i;
     for(i=0;i<FINEANGLES/8;i++)
     {
-        double tang=tan((i+0.5)/radtoint);
+        float tang=tanf((i+0.5f)/radtoint);
         finetangent[i]=(int32_t)(tang*GLOBAL1);
         finetangent[FINEANGLES/4-1-i]=(int32_t)((1/tang)*GLOBAL1);
     }
@@ -733,7 +733,7 @@
     float anglestep=(float)(PI/2/ANGLEQUAD);
     for(i=0; i<ANGLEQUAD; i++)
     {
-        fixed value=(int32_t)(GLOBAL1*sin(angle));
+        fixed value=(int32_t)(GLOBAL1*sinf(angle));
         sintable[i]=sintable[i+ANGLES]=sintable[ANGLES/2-i]=value;
         sintable[ANGLES-i]=sintable[ANGLES/2+i]=-value;
         angle+=anglestep;
@@ -764,9 +764,9 @@
     int     i;
     int    intang;
     float   angle;
-    double  tang;
+    float   tang;
     int     halfview;
-    double  facedist;
+    float   facedist;
 
     focallength = focal;
     facedist = focal+MINDIST;
@@ -792,7 +792,7 @@
     {
         // start 1/2 pixel over, so viewangle bisects two middle pixels
         tang = (int32_t)i*VIEWGLOBAL/viewwidth/facedist;
-        angle = (float) atan(tang);
+        angle = atanf(tang);
         intang = (int) (angle*radtoint);
         pixelangle[halfview-1-i] = intang;
         pixelangle[halfview+i] = -intang;
diff --git a/nds/opl.c b/nds/opl.c
--- a/nds/opl.c
+++ b/nds/opl.c
@@ -112,7 +112,7 @@
 
     for (i = 0; i < SIN_LEN; ++i)
     {
-        s16 s = (s16) lrint (sin ((i * 2 + 1) * M_PI / SIN_LEN) * 4095);
+        s16 s = (s16) lrintf (sinf ((i * 2 + 1) * (float) M_PI / SIN_LEN) * 4095);
         int quarter = i / (SIN_LEN / 4);
 
         sin_tab[0][i] = s;
@@ -121,7 +121,7 @@
         sin_tab[3][i] = quarter & 1 ? 0 : sin_tab[2][i];
     }
     for (w = 0; w < 32; ++w)
-        att_volume[w] = (u16) lrint (2047 * pow (2.0, -w / 32.0));
+        att_volume[w] = (u16) lrintf (2047 * powf (2.0f, -w / 32.0f));
 
     // The first 16 rates are infinite, rates 0-12 step every 2^(12-n)
     // samples and 13-15 every sample, with the increments of the MAME
@@ -129,7 +129,7 @@
     for (i = 0; i < 96; ++i)
     {
         int rate = (i - 16) / 4, sel, shift = 0;
-        double avg = 0, updates;
+        float avg = 0, updates;
 
         if (i < 16)
             sel = 14;
@@ -140,12 +140,12 @@
         }
         else
             sel = rate < 15 ? 4 + (rate - 13) * 4 + ((i - 16) & 3) : 12;
-        updates = (double) EG_QUANTUM / (1 << shift);
+        updates = (float) EG_QUANTUM / (1 << shift);
 
         for (w = 0; w < 8; ++w)
-            avg += eg_inc[sel][w] / 8.0;
+            avg += eg_inc[sel][w] / 8.0f;
         rate_inc[i] = (s32) (avg * updates * 65536);
-        rate_att[i] = (u32) (pow (1.0 - avg / 8.0, updates) * 65536);
+        rate_att[i] = (u32) (powf (1.0f - avg / 8.0f, updates) * 65536);
     }
 }
 
diff --git a/wl_utils.h b/wl_utils.h
--- a/wl_utils.h
+++ b/wl_utils.h
@@ -11,6 +11,7 @@
 void     *safe_malloc (size_t size, const char *fname, uint32_t line);
 fixed    FixedMul (fixed a, fixed b);
 fixed    FixedDiv (fixed a, fixed b);
+int      VectorAngle (int32_t y, int32_t x);
 
 uint16_t ReadShort (void *ptr);
 uint32_t ReadLong (void *ptr);
diff --git a/wl_utils.c b/wl_utils.c
--- a/wl_utils.c
+++ b/wl_utils.c
@@ -38,6 +38,18 @@
 	return div64((int64_t)a << FRACBITS, b);
 }
 
+// Convert a direction to the game's integer angles without double math.
+int VectorAngle (int32_t y, int32_t x)
+{
+    float angle = atan2f ((float) y, (float) x);
+    int result;
+
+    if (angle < 0)
+        angle += 2.0f * PI;
+    result = (int) (angle * (ANGLES / (2.0f * PI)));
+    return result == ANGLES ? 0 : result;
+}
+
 uint16_t ReadShort (void *ptr)
 {
     unsigned value;
diff --git a/wl_act2.c b/wl_act2.c
--- a/wl_act2.c
+++ b/wl_act2.c
@@ -1639,15 +1639,11 @@
 {
     objtype *newobj = NULL;
     int32_t deltax,deltay;
-    float   angle;
     int     iangle;
 
     deltax = player->x - ob->x;
     deltay = ob->y - player->y;
-    angle = (float) atan2 ((float) deltay, (float) deltax);
-    if (angle<0)
-        angle = (float) (M_PI*2+angle);
-    iangle = (int) (angle/(M_PI*2)*ANGLES);
+    iangle = VectorAngle (deltay, deltax);
     if (ob->obclass == deathobj)
     {
         T_Shoot (ob);
@@ -2311,15 +2307,11 @@
 {
     objtype *newobj = NULL;
     int32_t deltax,deltay;
-    float   angle;
     int     iangle;
 
     deltax = player->x - ob->x;
     deltay = ob->y - player->y;
-    angle = (float) atan2((float) deltay, (float) deltax);
-    if (angle<0)
-        angle = (float) (M_PI*2+angle);
-    iangle = (int) (angle/(M_PI*2)*ANGLES);
+    iangle = VectorAngle (deltay, deltax);
 
     newobj = GetNewActor();
     newobj->state = &s_needle1;
@@ -2352,15 +2344,11 @@
 {
     objtype *newobj = NULL;
     int32_t deltax,deltay;
-    float   angle;
     int     iangle;
 
     deltax = player->x - ob->x;
     deltay = ob->y - player->y;
-    angle = (float) atan2((float) deltay, (float) deltax);
-    if (angle<0)
-        angle = (float) (M_PI*2+angle);
-    iangle = (int) (angle/(M_PI*2)*ANGLES);
+    iangle = VectorAngle (deltay, deltax);
 
     newobj = GetNewActor();
     newobj->state = &s_rocket;
@@ -2945,7 +2933,6 @@
 {
     objtype *newobj = NULL;
     int32_t deltax,deltay;
-    float   angle;
     int     iangle;
 
     if (!objfreelist)       // stop shooting if over MAXACTORS
@@ -2956,10 +2943,7 @@
 
     deltax = player->x - ob->x;
     deltay = ob->y - player->y;
-    angle = (float) atan2((float) deltay, (float) deltax);
-    if (angle<0)
-        angle = (float)(M_PI*2+angle);
-    iangle = (int) (angle/(M_PI*2)*ANGLES);
+    iangle = VectorAngle (deltay, deltax);
 
     newobj = GetNewActor();
     newobj->state = &s_fire1;
@@ -3804,7 +3788,6 @@
 void    A_StartDeathCam (objtype *ob)
 {
     int32_t dx,dy;
-    float   fangle;
     int32_t xmove,ymove;
     int32_t dist;
 
@@ -3857,11 +3840,7 @@
     dx = ob->x - player->x;
     dy = player->y - ob->y;
 
-    fangle = (float) atan2((float) dy, (float) dx);          // returns -pi to pi
-    if (fangle<0)
-        fangle = (float) (M_PI*2+fangle);
-
-    player->angle = (short) (fangle/(M_PI*2)*ANGLES);
+    player->angle = VectorAngle (dy, dx);
 
     //
     // try to position as close as possible without being in a wall
diff --git a/wl_game.c b/wl_game.c
--- a/wl_game.c
+++ b/wl_game.c
@@ -1197,7 +1197,6 @@
 
 void Died (void)
 {
-    float   fangle;
     int32_t dx,dy;
     int     iangle,curangle,clockwise,counter,change;
 
@@ -1218,11 +1217,7 @@
         dx = LastAttacker->x - player->x;
         dy = player->y - LastAttacker->y;
 
-        fangle = (float) atan2((float) dy, (float) dx);     // returns -pi to pi
-        if (fangle<0)
-            fangle = (float) (M_PI*2+fangle);
-
-        iangle = (int) (fangle/(M_PI*2)*ANGLES);
+        iangle = VectorAngle (dy, dx);
     }
     else
     {
diff --git a/wl_inter.c b/wl_inter.c
--- a/wl_inter.c
+++ b/wl_inter.c
@@ -424,22 +424,22 @@
         //
         // Episode One Par Times
         //
-        {1.5, "01:30"},
+        {1.5f, "01:30"},
         {2, "02:00"},
         {2, "02:00"},
-        {3.5, "03:30"},
+        {3.5f, "03:30"},
         {3, "03:00"},
         {3, "03:00"},
-        {2.5, "02:30"},
-        {2.5, "02:30"},
+        {2.5f, "02:30"},
+        {2.5f, "02:30"},
         {0, "??:??"},           // Boss level
         {0, "??:??"},           // Secret level
 
         //
         // Episode Two Par Times
         //
-        {1.5, "01:30"},
-        {3.5, "03:30"},
+        {1.5f, "01:30"},
+        {3.5f, "03:30"},
         {3, "03:00"},
         {2, "02:00"},
         {4, "04:00"},
@@ -452,12 +452,12 @@
         //
         // Episode Three Par Times
         //
-        {1.5, "01:30"},
-        {1.5, "01:30"},
-        {2.5, "02:30"},
-        {2.5, "02:30"},
-        {3.5, "03:30"},
-        {2.5, "02:30"},
+        {1.5f, "01:30"},
+        {1.5f, "01:30"},
+        {2.5f, "02:30"},
+        {2.5f, "02:30"},
+        {3.5f, "03:30"},
+        {2.5f, "02:30"},
         {2, "02:00"},
         {6, "06:00"},
         {0, "??:??"},
@@ -468,60 +468,60 @@
         //
         {2, "02:00"},
         {2, "02:00"},
-        {1.5, "01:30"},
+        {1.5f, "01:30"},
         {1, "01:00"},
-        {4.5, "04:30"},
-        {3.5, "03:30"},
+        {4.5f, "04:30"},
+        {3.5f, "03:30"},
         {2, "02:00"},
-        {4.5, "04:30"},
+        {4.5f, "04:30"},
         {0, "??:??"},
         {0, "??:??"},
 
         //
         // Episode Five Par Times
         //
-        {2.5, "02:30"},
-        {1.5, "01:30"},
-        {2.5, "02:30"},
-        {2.5, "02:30"},
+        {2.5f, "02:30"},
+        {1.5f, "01:30"},
+        {2.5f, "02:30"},
+        {2.5f, "02:30"},
         {4, "04:00"},
         {3, "03:00"},
-        {4.5, "04:30"},
-        {3.5, "03:30"},
+        {4.5f, "04:30"},
+        {3.5f, "03:30"},
         {0, "??:??"},
         {0, "??:??"},
 
         //
         // Episode Six Par Times
         //
-        {6.5, "06:30"},
+        {6.5f, "06:30"},
         {4, "04:00"},
-        {4.5, "04:30"},
+        {4.5f, "04:30"},
         {6, "06:00"},
         {5, "05:00"},
-        {5.5, "05:30"},
-        {5.5, "05:30"},
-        {8.5, "08:30"},
+        {5.5f, "05:30"},
+        {5.5f, "05:30"},
+        {8.5f, "08:30"},
         {0, "??:??"},
         {0, "??:??"}
 #else
         //
         // SPEAR OF DESTINY TIMES
         //
-        {1.5, "01:30"},
-        {3.5, "03:30"},
-        {2.75, "02:45"},
-        {3.5, "03:30"},
+        {1.5f, "01:30"},
+        {3.5f, "03:30"},
+        {2.75f, "02:45"},
+        {3.5f, "03:30"},
         {0, "??:??"},           // Boss 1
-        {4.5, "04:30"},
-        {3.25, "03:15"},
-        {2.75, "02:45"},
-        {4.75, "04:45"},
+        {4.5f, "04:30"},
+        {3.25f, "03:15"},
+        {2.75f, "02:45"},
+        {4.75f, "04:45"},
         {0, "??:??"},           // Boss 2
-        {6.5, "06:30"},
-        {4.5, "04:30"},
-        {2.75, "02:45"},
-        {4.5, "04:30"},
+        {6.5f, "06:30"},
+        {4.5f, "04:30"},
+        {2.75f, "02:45"},
+        {4.5f, "04:30"},
         {6, "06:00"},
         {0, "??:??"},           // Boss 3
         {6, "06:00"},
